signac/core/caching.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implement the caching feature to  SyncedCollection API."""
+import uuid
+import logging
+import pickle
+from collections.abc import MutableMapping
+
+logger = logging.getLogger(__name__)
+
+
+def get_cache():
+    """Return the cache.
+
+    This method returns an instance of :class:`~RedisCache` if redis-server is available,
+    or otherwise an instance of ``dict`` for an in-memory cache.
+
+    Returns
+    -------
+    cache
+        An instance of :class:`~RedisCache` if redis-server is available, otherwise dictionary.
+    """
+    try:
+        import redis
+        REDIS = True
+    except ImportError as error:
+        logger.debug(str(error))
+        REDIS = False
+    if REDIS:
+        try:
+            # try to connect to server
+            cache = redis.Redis()
+            test_key = str(uuid.uuid4())
+            cache.set(test_key, 0)
+            assert cache.get(test_key) == b'0'  # Redis store data as bytes
+            cache.delete(test_key)
+            logger.info("Using Redis cache.")
+            return RedisCache(cache)
+        except (redis.exceptions.ConnectionError, AssertionError) as error:
+            logger.debug(str(error))
+    logger.info("Redis not available.")
+    return {}
+
+
+class RedisCache(MutableMapping):
+    """Redis backend based cache.
+
+    The RedisCache is a :class:`~collections.abc.MutableMapping`. It uses redis-server
+    to store data by using redis client. Redis client only accepts data as bytes, strings
+    or numbers (ints, longs and floats) and returns response as bytes. So, this uses pickle
+    serialization to convert data into strings.
+    """
+
+    def __init__(self, client):
+        self._client = client
+
+    def __setitem__(self, key, value):
+        self._client[key] = pickle.dumps(value)
+
+    def __getitem__(self, key):
+        return pickle.loads(self._client[key])
```

Comment:
Is pickling necessary here? What do we need feature-wise that isn't natively supported by Redis? I believe pickling is somewhat slow, so if we can just use JSON-encoded strings and other supported primitives, that would probably be best.

Reply:
I don't think it's as much a question of what Redis doesn't support as a question of what objects can be converted into bytes via JSON. The question is how we want to restrict the input data. See my main comment on the conversation for a more detailed discussion.

Reply:
@vyasr In my understanding, the cache (Redis or dict) stores the contents of entire files (i.e. valid JSON strings, represented as bytes). If that's correct, we should never need to pickle and we shouldn't have any issues with object types -- we only use byte-strings of JSON-encoded data, right? Please correct me if I'm missing something.

(Reminder to update the docstring to match whatever changes are made here, or make no claims about serialization in the docstring.)

Reply:
What you're missing is that SyncedCollections are not intrinsically restricted to JSON-encodeable data. That restriction is only made at the level of validators, and for using these classes in signac the plan is to employ JSON validation for all collections. That's also why I created the `_encode` and `_decode` hooks in the `FileBufferedSyncedCollection` class, so that other choices could be made in subclasses if necessary. I think the most extensible solution for this is to initialize a Redis cache with pointers to the encode/decode methods of the associated collection (which we would probably want to be staticmethods in that case), that way the same data restrictions would apply in both cases and different buffering classes could configure it.

Since we're holding off on merging the Redis caching anyway, the easiest solution for now is to have the `FileBufferedSyncedCollection` default to JSON encoding and document that as a restriction, then come back to it at a later point. I don't know of any better alternatives at the moment, and since buffering is an internal implementation detail that we only require to support signac at the moment, it's not a big deal. I think the hooks I've added should be sufficient to enable alternative encodings easily if necessary.

Reply:
That's fine as a way to move this forward. Generally I think it might be fine to assume that any cache that can store "bytes" is sufficient, which would cover data formats beyond JSON and would apply to most forms of cache (`dict`, Redis, ...).

Reply:
Another reason that I 'm pretty confident that the cache _must_ store its values as `bytes`: that's also an expectation of the hashing function.